x-pack/plugin/ccr/src/main/java/org/elasticsearch/xpack/ccr/repository/CcrRestoreSourceService.java

```diff
@@ -0,0 +1,132 @@
+
+    private static final Logger logger = LogManager.getLogger(CcrRestoreSourceService.class);
+
+    private final Map<String, Engine.IndexCommitRef> onGoingRestores = ConcurrentCollections.newConcurrentMap();
+    private final Map<IndexShard, HashSet<String>> sessionsForShard = new HashMap<>();
+    private final CopyOnWriteArrayList<Consumer<String>> openSessionListeners = new CopyOnWriteArrayList<>();
+    private final CopyOnWriteArrayList<Consumer<String>> closeSessionListeners = new CopyOnWriteArrayList<>();
+
+    public CcrRestoreSourceService(Settings settings) {
+        super(settings);
+    }
+
+    // TODO: Need to register with IndicesService
+    @Override
+    public synchronized void beforeIndexShardClosed(ShardId shardId, @Nullable IndexShard indexShard, Settings indexSettings) {
+        if (indexShard != null) {
+            logger.debug("shard [{}] closing, closing sessions", indexShard);
+            HashSet<String> sessions = sessionsForShard.remove(indexShard);
+            if (sessions != null) {
+                for (String sessionUUID : sessions) {
+                    logger.debug("closing session [{}] for shard [{}]", sessionUUID, indexShard);
+                    Engine.IndexCommitRef commit = onGoingRestores.remove(sessionUUID);
+                    IOUtils.closeWhileHandlingException(commit);
+                }
+            }
+        }
+    }
+
+    @Override
+    protected void doStart() {
+
+    }
+
+    @Override
+    protected void doStop() {
+
+    }
+
+    @Override
+    protected synchronized void doClose() throws IOException {
+        sessionsForShard.clear();
+        IOUtils.closeWhileHandlingException(onGoingRestores.values());
+        onGoingRestores.clear();
+    }
+
+    // TODO: The listeners are for testing. Once end-to-end file restore is implemented and can be tested,
+    //  these should be removed.
+    public void addOpenSessionListener(Consumer<String> listener) {
+        openSessionListeners.add(listener);
+    }
+
+    public void addCloseSessionListener(Consumer<String> listener) {
+        closeSessionListeners.add(listener);
+    }
+
+    // default visibility for testing
+    synchronized HashSet<String> getSessionsForShard(IndexShard indexShard) {
+        return sessionsForShard.get(indexShard);
+    }
+
+    // default visibility for testing
+    synchronized Engine.IndexCommitRef getIndexCommit(String sessionUUID) {
+        return onGoingRestores.get(sessionUUID);
+    }
+
+    public synchronized Store.MetadataSnapshot openSession(String sessionUUID, IndexShard indexShard) throws IOException {
+        logger.debug("opening session [{}] for shard [{}]", sessionUUID, indexShard);
+        Engine.IndexCommitRef commit;
+        if (onGoingRestores.containsKey(sessionUUID)) {
+            logger.debug("session [{}] already exists", sessionUUID);
+            commit = onGoingRestores.get(sessionUUID);
```

Comment:
should we be so lenient here, or rather reject opening a session which is already supposed to exist? It depends on how we want to handle failures / retries

Reply:
I would kind of prefer the `put` be idempotent. This is also why I have the session uuid is generated on the follower node. I'll explain more about my design in a top-level comment. Maybe we should add validation to prevent (unlikely) uuid conflicts (ensure that the put session request comes from the same follower node)?